api/utils/keys/hardwarekey/cachingprompt_test.go

```diff
@@ -0,0 +1,110 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package hardwarekey defines types and interfaces for hardware private keys.
+
+package hardwarekey_test
+
+import (
+	"context"
+	"math/rand"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/gravitational/teleport/api/utils/keys/hardwarekey"
+)
+
+func TestPINCachingPrompt(t *testing.T) {
+	ctx := context.Background()
+
+	// Note: locally this test gets flaky around 10µs.
+	cacheTTL := 10 * time.Millisecond
+	cachingPrompt := hardwarekey.NewPINCachingPrompt(&randPINPrompt{}, cacheTTL)
+
+	t.Run("AskPIN", func(t *testing.T) {
+		// prompt and cache a new PIN for 100ms.
+		cachedPIN, err := cachingPrompt.AskPIN(ctx, hardwarekey.PINRequired, hardwarekey.ContextualKeyInfo{})
+		require.NoError(t, err)
+		timer := time.NewTimer(cacheTTL)
+
+		// Check that the PIN remains cached.
+		for i := 0; i < 3; i++ {
+			pin, err := cachingPrompt.AskPIN(ctx, hardwarekey.PINRequired, hardwarekey.ContextualKeyInfo{})
+			require.NoError(t, err)
+			require.Equal(t, cachedPIN, pin)
+		}
+
+		// Check that the PIN is not cached after 100ms.
+		<-timer.C
+		pin, err := cachingPrompt.AskPIN(ctx, hardwarekey.PINRequired, hardwarekey.ContextualKeyInfo{})
+		require.NoError(t, err)
+		require.NotEqual(t, cachedPIN, pin)
+	})
+
+	t.Run("ChangePIN", func(t *testing.T) {
+		// ChangePIN should prompt and cache a new PIN for 100ms.
+		pinAndPUK, err := cachingPrompt.ChangePIN(ctx, hardwarekey.ContextualKeyInfo{})
+		require.NoError(t, err)
+		cachedPIN := pinAndPUK.PIN
+		timer := time.NewTimer(cacheTTL)
+
+		// Check that the PIN remains cached.
+		for i := 0; i < 3; i++ {
+			pin, err := cachingPrompt.AskPIN(ctx, hardwarekey.PINRequired, hardwarekey.ContextualKeyInfo{})
+			require.NoError(t, err)
+			require.Equal(t, cachedPIN, pin)
+		}
+
+		// Check that the PIN is not cached after 100ms.
+		<-timer.C
+		pin, err := cachingPrompt.AskPIN(ctx, hardwarekey.PINRequired, hardwarekey.ContextualKeyInfo{})
+		require.NoError(t, err)
+		require.NotEqual(t, cachedPIN, pin)
+	})
+
+}
+
+type randPINPrompt struct {
+	lastPin int
+}
+
+func (p *randPINPrompt) AskPIN(ctx context.Context, requirement hardwarekey.PINPromptRequirement, keyInfo hardwarekey.ContextualKeyInfo) (string, error) {
+	return p.randPIN(), nil
+}
+
+func (p *randPINPrompt) Touch(ctx context.Context, keyInfo hardwarekey.ContextualKeyInfo) error {
+	return nil
+}
+
+func (p *randPINPrompt) ChangePIN(ctx context.Context, keyInfo hardwarekey.ContextualKeyInfo) (*hardwarekey.PINAndPUK, error) {
+	return &hardwarekey.PINAndPUK{
+		PIN: p.randPIN(),
+	}, nil
+}
+
+func (p *randPINPrompt) ConfirmSlotOverwrite(ctx context.Context, message string, keyInfo hardwarekey.ContextualKeyInfo) (bool, error) {
+	return false, nil
+}
+
+func (p *randPINPrompt) randPIN() string {
```

Comment:
Isn't `fmt.Sprintf("%08d", rand.Intn(100000000))` sufficient?

Reply:
You mean instead of `strconv.Itoa`? `strconv.Itoa` seems cleaner imo, but if there's a reason to prefer `fmt.Sprintf` then we can do that.

If you mean instead of having the `randPIN` method, that's to ensure we don't repeat the same random PIN. Trying to avoid the 0.0000011% chance of a flaky test 😉

Reply:
It seems the ` + 100000000` thing you're doing is to avoid leading zeroes? `Sprintf` will take care of that while simultaneously allowing you to use the full range of valid PINs. I don't really care about the 1 in a billion chance of a flaky test. We'll encounter 500 other flakes before we ever see this one!